include/dali/core/fast_div.h

```diff
@@ -0,0 +1,255 @@
+  lohi<T> ret;
+  ret.lo = a.lo - b.lo;
+  int borrow = (b.lo && ret.lo > a.lo);
+  ret.hi = a.hi - b.hi - borrow;
+  return ret;
+}
+
+__host__ __device__ inline uint32_t div_lohi(uint32_t lo, uint32_t hi, uint32_t divisor) {
+  return (static_cast<uint64_t>(hi) << 32 | lo) / divisor;
+}
+
+__host__ __device__ __forceinline__ lohi<uint64_t> mull(uint64_t a, uint64_t b) {
+  lohi<uint64_t> ret;
+#ifdef __CUDA_ARCH__
+  ret.lo = a * b;
+  ret.hi = __umul64hi(a, b);
+#else
+  unsigned __int128 m = (unsigned __int128)a * b;
+  ret.lo = m;
+  ret.hi = m >> 64;
+#endif
+  return ret;
+}
+
+template <typename uint>
+__host__ __device__ int ilog2(uint x) noexcept {
+  int n = 0;
+  while (x >>= 1)
+    n++;
+  return n;
+};
+
+__host__ __device__ inline uint64_t div_lohi(uint64_t lo, uint64_t hi, uint64_t divisor) {
+#if defined(__x86_64) && defined(__GNUC__) && !defined(__CUDA_ARCH__)
+  // I hope this gets compiled to dividing rdx:rax register pair by a 64-bit value
+  return (static_cast<unsigned __int128>(hi) << 64 | lo) / divisor;
+#else
+  // NVCC doesn't support __int128 in device code, so we need a bit of hackery
+  if (!hi)  // No high part? Just divide in 64-bits and be done.
+    return lo / divisor;
+
+  // long division:
+
+  int lnum = ilog2(hi) + 64;
+  int lden = ilog2(divisor);
+
+  lohi<uint64_t> num = { lo, hi };
+  lohi<uint64_t> den = { divisor, 0 };
+
+  // calculate MSB positions...
+  int sh = lnum - lden;
+  // .. and align numerator and denominator
+  den <<= sh;
+
+  uint64_t q = 0;
+
+  while (sh >= 0) {
+    lohi<uint64_t> dif = num - den;  // this serves both as difference and comparison
+    if (static_cast<int64_t>(dif.hi) >= 0) {
+      num = dif;
+      q |= static_cast<uint64_t>(1) << sh;
+    }
+    sh--;
+    den >>= 1;
+  }
+  return q;
+#endif
+}
+
+}  // namespace detail
+
+/**
+ * @brief Fast unsigned integer division
+ *
+ * This class is useful when there's a need calculate division (and/or remainder) of many numbers
+ * by common divisor, not known at compile time.
+ *
+ * Based on:
+ * Labor of Division (Episode III): Faster Unsigned Division by Constants
+ * ridiculous_fish corydoras@ridiculousfish.com
+ * https://ridiculousfish.com/blog/posts/labor-of-division-episode-iii.html
+ *
+ * Limitations:
+ * - doesn't work for maximum value of uint when dividing by "uncooperative" numbers
+ */
+template <typename uint>
+struct fast_div {
+  uint divisor;
+  uint mul;
+  uint8_t add;
+  uint8_t shift;
+
+  static_assert(std::is_integral<uint>::value && std::is_unsigned<uint>::value,
+                "fast_div works only with unsigned integers");
+
+  __host__ __device__ fast_div() {}
+
+  __host__ __device__ fast_div(uint divisor) {  // NOLINT
+    init(divisor);
+  }
+
+  __host__ __device__ void init(uint divisor) {
+    this->divisor = divisor;
+    this->mul = 1;
+    this->shift = 0;
+    this->add = 0;
+    if (divisor == 0) {
+      return;
+    }
+    int l = detail::ilog2(divisor);
```

Comment:
Descriptive name for this would make it easier for me to read it, especially in scenario where `l` and `1` look similar.

Reply:
Yeah... and there's an unnecessary "try harder" assignment a few lines below. I'll change that.